### PR TITLE
Add spec for dex

### PR DESCRIPTION
**Pull request by codchen**

### CODE REVIEW

File: x/dex/spec/README.md [hunk 1]
```diff
@@ -1,21 +1,46 @@
...
+`dex` module is responsible for matching orders for registered contracts.
 
 ## Concepts
-
-## State
-
-## Messages
-
-## Events
+### Frequent Batch Auction
```
Comment (jay-seilabs) [inline x/dex/spec/README.md:5]: can we add ordering of how orders are processed, similar to whitepaper (cancels first, then add limit orders, then market orders, then fill limit orders)

File: x/dex/spec/README.md [hunk 2]
```diff
@@ -1,21 +1,46 @@
...
+Since `dex` only provides order matching logic, product logic specific to individual protocols still needs to be defined in CosmWasm contracts. As such, `dex` offers a way to inform the protocol contracts about order placement and matching results. `dex` achieves this by requiring contracts that want to leverage `dex`'s order matching logic to explicitly register via a special transaction type `MsgRegisterContract`.
+
+## State (KV Store)
+The following prefixes are persisted in disk:
+- "LongBook-value-": order book state on the long side where each entry represents a price level and can contain multiple orders at that same price.
+- "ShortBook-value-": similar to the above but on the short side.
+- "x-wasm-contract": contract registration information.
+- "MatchResult-": match results of the most recent block.
+
+The following prefixes are only used intrablock and are cleared before committing the block:
```
Comment (jay-seilabs) [inline x/dex/spec/README.md]: can you give some more context on why these prefixes are needed (i.e. what problem we ran into that resulted in us using this approach)

File: x/dex/spec/README.md [hunk 3]
```diff
@@ -1,21 +1,46 @@
...
+- "MemOrder-": orders added by transactions in the current block and will be matched against the order book states at the end of the block.
+- "MemCancel-": cancellations added by transactions in the current block and will update the order book states at the end of the block.
 
 ## Hooks
```
Comment (jay-seilabs) [inline x/dex/spec/README.md:39]: can we briefly explain how we prevent negative behavior from developers here (the gas logic)

**Review by jay-seilabs — APPROVED** [review]

File: x/dex/spec/README.md [hunk 4]
```diff
@@ -1,21 +1,46 @@
...
-
-## Events
+### Frequent Batch Auction
+The traditional implementation of exchange logic would look something like the following:
+1. User A sends an order placement transaction
+2. User B sends another order placement transaction
+3. User A's transaction is processed by matching it against the order book state and settle accordingly
+4. User B's transaction is processed similarly
+
+Sei's `dex` module takes a different approach:
```
Comment (LCyson) [inline x/dex/spec/README.md:12]: can we specify the difference between limit order processing and market order processing here? probably elaborates the step 1-5 here would be good: https://github.com/sei-protocol/sei-chain/blob/master/x/dex/contract/execution.go#L59-L74

File: x/dex/spec/README.md [hunk 5]
```diff
@@ -1,21 +1,46 @@
...
+1. User A sends an order placement transaction
+2. User B sends another order placement transaction
+3. User A's transaction is processed by simply adding the order to an in-memory queue, without matching
+4. User B's transaction is processed similarly
+5. At the end of the block that contains both transactions, the in-memory queue as a whole will be matched against the order book state
+
+### Contract Registration
+Since `dex` only provides order matching logic, product logic specific to individual protocols still needs to be defined in CosmWasm contracts. As such, `dex` offers a way to inform the protocol contracts about order placement and matching results. `dex` achieves this by requiring contracts that want to leverage `dex`'s order matching logic to explicitly register via a special transaction type `MsgRegisterContract`.
```
Comment (LCyson) [inline x/dex/spec/README.md:26]: can we elaborate what specifically can be registered here? i.e. tick size, quantity size and dependencies

Comment (LCyson) [reply x/dex/spec/README.md:26]: and where those variables are registered/accessed + how they are used

File: x/dex/spec/README.md [hunk 6]
```diff
@@ -1,21 +1,46 @@
...
+
+## State (KV Store)
```
Comment (LCyson) [inline x/dex/spec/README.md:28]: can we briefly explain what are memState (right now KVStore wrapper) and the dirty orderbook, and how they're used in the step 1-5 above

Comment (LCyson) [reply x/dex/spec/README.md:28]: Also we may want to talk about how multiple contracts are safely processed concurrently with these data structures

File: x/dex/spec/README.md [hunk 7]
```diff
@@ -1,21 +1,46 @@
...
 ## Hooks
+A registered contract can define the following `sudo` hooks that will be called by the `dex` module at appropriate times:
+- BulkOrderPlacements: informs the contract about order placements
+- BulkOrderCancellations: informs the contract about order cancellations
+- Settlement: informs the contract about matched orders and the settlement prices
 
-## Parameters
+There are two more utility hooks that a contract can define for housekeeping purposes (e.g. recalculate TWAPs):
+- NewBlock
+- FinalizeBlock (note that this is distinct from ABCI++'s FinalizeBlock)
```
Comment (LCyson) [inline x/dex/spec/README.md:47]: can we briefly talk about what variables/responses will be passed through this sudo call

**Pull request merged**